packages/router/src/adapters/contract/contract.ts

```diff
@@ -44,6 +45,197 @@ export const removeLiquidityEvt = new Evt<{
 export const isChainSupportedByGelato = _isChainSupportedByGelato;
 export const gelatoSend = _gelatoSend;
 
+/**
+ * Helper method for waiting for relayer transaction go through using event handling. Retrieves the receipt afterwards
+ * and increments the amount spent on relayer fees.
+ *
+ * @param chainId - Chain ID of the transaction.
+ * @param transactionId - Transaction ID of the transaction to wait for.
+ * @param evt - Event to wait for.
+ * @param relayerFee - Relayer fee amount.
+ * @param relayerFeeAsset - Relayer fee asset ID.
+ * @param reason - Reason for the transaction (i.e. transaction type).
+ * @param requestContext - Request context.
+ * @returns TransactionReceipt of the transaction (if it's successful).
+ * @throws Timeout error if the event is not emitted within the timeout period.
+ */
+const waitForRelayer = async (
+  chainId: number,
+  transactionId: string | undefined,
+  evt: Evt<any>,
+  relayerFee: {
+    asset: string;
+    amount: string;
+  },
+  reason: TransactionReason,
+  requestContext: RequestContext,
+) => {
+  const { txService } = getContext();
+  // Listen for event on contract.
+  const { event } = transactionId
+    ? await evt.pipe(({ args }) => args.txData.transactionId === transactionId).waitFor(300_000)
+    : await evt.waitFor(300_000);
+  // Retrieve the receipt.
+  const receipt = await txService.getTransactionReceipt(chainId, event.transactionHash);
```

Comment:
Is there a case where this would return `undefined`? The [function](https://github.com/connext/nxtp/blob/153dbe146ea52ec5f583407630d9f9e79ab19ea0/packages/txservice/src/chainreader.ts#L158-L160) doesn't have that option in the types, but it is in the [ethers docs](https://docs.ethers.io/v5/api/providers/provider/#Provider-getTransactionReceipt).

We should probably update the types in the `chainreader` and handle the `undefined` case here.